pkg/kgo/config.go

```diff
@@ -1317,6 +1319,23 @@ func RequireStableFetchOffsets() GroupOpt {
 	return groupOpt{func(cfg *cfg) { cfg.requireStable = true }}
 }
 
+// AdjustOffsetsBeforeAssign sets the function to be called when a group is joined
+// after offsets are fetched for those partitions so that a user can adjust them
+// before consumption begins.
+//
+// This function combined with OnPartitionsRevoked should not exceed the
+// rebalance interval. It is possible for the group, immediately after
+// finishing a balance, to re-enter a new balancing session.
+//
+// The AdjustOffsetsBeforeAssign function is passed the client's context, which is
+// only canceled if the client is closed.
+//
+// This function is not called concurrent with any other On callback, and this
+// function is given a new map that the user is free to modify.
```

Comment:
This currently races with the callback for `OnPartitionsAssigned`, since fetching offsets is done in a separate goroutine. However, the purpose of this function isn't to modify anything that is passed to `OnPartitionsAssigned`, it's to modify the actual offsets that a consumer is about to consume. As well, this is passed a pretty distinct context: the context used here is canceled after OnRevoked finishes (after the heartbeat loop returns). As well, this doesn't duplicate the input map, but I think that's fine in this case. Also there's one more footgun that a user has to be aware of: epoch validation.

I think the doc bit above can be updated to:
```go
// AdjustOffsetsBeforeAssign sets the function to be called when a group is joined
// after offsets are fetched for those partitions so that a user can adjust them
// before consumption begins.
//
// This function combined should not exceed the rebalance interval. It is possible
// for the group, immediately after finishing a balance, to re-enter a new balancing
// session. This function is passed a context that is canceled if the current group
// session finishes (i.e., after revoking).
//
// If you are resetting the position of the offset, you may want to clear any existing
// "epoch" with WithEpoch(-1). If the epoch is non-negative, the client performs
// data loss detection, which may result in errors and unexpected consuming offsets.
```

Also, what do you think about `AdjustOffsetsBeforeAssignFn`? It's a bit of an ugly name but it somewhat fits a pattern: minus `Dialer`, the existing options that accept functions are all either `On`, `Callback`, or `Fn`:
```go
func Dialer(fn func(ctx context.Context, network, host string) (net.Conn, error)) Opt
func RetryBackoffFn(backoff func(int) time.Duration) Opt
func RetryTimeoutFn(t func(int16) time.Duration) Opt 
func ProducerOnDataLossDetected(fn func(string, int32)) ProducerOpt
func AutoCommitCallback(fn func(*Client, *kmsg.OffsetCommitRequest, *kmsg.OffsetCommitResponse, error)) GroupOpt
func OnPartitionsAssigned(onAssigned func(context.Context, *Client, map[string][]int32)) GroupOpt
func OnPartitionsLost(onLost func(context.Context, *Client, map[string][]int32)) GroupOpt
func OnPartitionsRevoked(onRevoked func(context.Context, *Client, map[string][]int32)) GroupOpt
```
"callback" could also work here but "fn" feels more like what this is doing. Maybe `AdjustFetchOffsetsFn`? I wonder if the current name could be confused with existing OnPartitionsAssigned semantics, when this is quite separate.

Reply:
Works for me - I’ll update the PR when I’m back at my keyboard 👍

Reply:
https://github.com/twmb/franz-go/pull/120/commits/2d6745063fbc8fe9cf678b21d0d0a7c8e8f4d93f

Reply:
Went with AdjustFetchOffsetsFn for brevity